types/src/language_storage.rs

```diff
@@ -22,7 +22,21 @@ pub struct StructTag {
     pub address: AccountAddress,
     pub module: Identifier,
     pub name: Identifier,
-    pub type_params: Vec<StructTag>,
+    pub type_params: Vec<TypeTag>,
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Hash, Eq, Clone, PartialOrd, Ord)]
+pub enum PrimitiveTypeTag {
+    Bool,
+    U64,
+    ByteArray,
+    Address,
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Hash, Eq, Clone, PartialOrd, Ord)]
+pub enum TypeTag {
+    Primitive(PrimitiveTypeTag),
```

Comment:
why not "inlining" the primitive into type tag? maybe that was part of the conversation that happened already?

Reply:
Not part of a previous conversation. I didn't inline the primitive variants as I knew we'll likely add a lot more (u8, u128, etc.) in the future. I might be worrying too much. Would love to hear the advantages & disadvantages of flattened & non-flattened enums.

Reply:
sorry I only saw this now :( my bad!
I am not sure what the worry is with adding more types. Can you explain?
Also I think this is changing given the latest work on the canonical ser/deser?
In any case it seems a TypeTag _could_ be a type tag and as such not have multiple level of indirection as in primitives and structs.
Anyway once you have this up again with the latest code we can talk a bit better... :)